### PR TITLE
[mlir][vector] Adds pattern rewrite for maskable Ops

### LINKED ISSUE #78787: [mlir] Special handling for pattern rewrite on maskable op
Pattern rewrite on maskable ops requires special handing when they are in `vector.mask` region as the `vector.mask` region only allows a single maskable op (https://discourse.llvm.org/t/rfc-vector-masking-representation-in-mlir/64964)

In https://github.com/llvm/llvm-project/pull/76468 we add some special checks in related pattern rewrites to skip those cases. But as the comment https://github.com/llvm/llvm-project/pull/76468#issuecomment-1897699116, we can provide a specialized pattern rewrite base class to help writing patterns for maskable op. For example, transform `vector.mask` and its inner maskable op together

### PR DESCRIPTION
**Pull request by banach-space**
Adds a generic pattern rewrite for maskable Ops, `MaskableOpRewritePattern`,
that will work for both masked and un-masked cases, e.g. for both:

* `vector.mask {vector.contract}` (masked), and
* `vector.contract` (not masked).

This helps to reduce code-duplication and standardise how we implement such
patterns.

Fixes #78787

### CODE REVIEW

**Review by hanhanW — APPROVED** [review]
This looks good to me, but perhaps we want some feedback from @dcaballe

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 1]
```diff
@@ -212,6 +211,64 @@ static Value createMul(Location loc, Value x, Value y, bool isInt,
...
+///   3. Invokes `matchAndRewriteMaskableOp` on Op that might be nested (not
+///   required) in the matched `vector.mask` operation from step 2.
+///
+/// It frees the patterns implementing this class from worrying about the
+/// logic to update the insertion point. However, those patterns are still
+/// responsible for providing an updated version of:
+///   * the source Op when mask _is not_ present,
+///   * the source Op *and* the mask Op when mask _is_ present.
+template <class SourceOp>
+struct MaybeMaskedOpRewritePattern : OpRewritePattern<SourceOp> {
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:229]: This is the main concern I've been struggling with for a while: this pattern matches the nested op and the rewrite would replace its parent. I'm not quite sure pattern rewrite allows that even though it works for this cases. We can make this match against the vector mask op but then we would need another pattern to match against the source op for the unmasked cases. I couldn't find an elegant way to provide this functionality but maybe you can :)

Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:229]: -> MaskableOpRewritePattern?

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 2]
```diff
@@ -212,6 +211,64 @@ static Value createMul(Location loc, Value x, Value y, bool isInt,
...
+  using OpRewritePattern<SourceOp>::OpRewritePattern;
+
+private:
+  LogicalResult matchAndRewrite(SourceOp sourceOp,
+                                PatternRewriter &rewriter) const final {
+    auto maskableOp =
+        dyn_cast_if_present<MaskableOpInterface>(sourceOp.getOperation());
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:236]: if_present?

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 3]
```diff
@@ -212,6 +211,64 @@ static Value createMul(Location loc, Value x, Value y, bool isInt,
...
+      maskOp = maskableOp.getMaskingOp();
+
+    // If this Op is masked, update the insertion point to avoid inserting into
+    // the vector.mask Op region.
+    OpBuilder::InsertionGuard guard(rewriter);
+    Operation *rootOp = sourceOp;
+    if (maskOp) {
+      rewriter.setInsertionPoint(maskOp);
+      rootOp = maskOp;
+    }
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:252]: I think you can fuse these two conditions

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 4]
```diff
@@ -212,6 +211,64 @@ static Value createMul(Location loc, Value x, Value y, bool isInt,
...
+  }
+
+public:
+  // Matches SourceOp that can potentially be masked with `maskingOp`. If the
+  // latter is present, returns an updated masking op (with a replacement for
+  // `sourceOp` nested inside). Otherwise, returns an updated `sourceOp`.
+  virtual FailureOr<Value>
+  matchAndRewriteMaskableOp(SourceOp sourceOp, MaskingOpInterface maskingOp,
+                            PatternRewriter &rewriter) const = 0;
+};
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:270]: Probably better to place this base class somewhere more public. Perhaps somewhere in vector utils or some other header?

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp [hunk 5]
```diff
@@ -634,52 +682,36 @@ struct UnrolledOuterProductGenerator
...
-  } else {
-    rootOp = op;
-  }
-
   UnrolledOuterProductGenerator e(rewriter, op);
   FailureOr<Value> matmatRes = e.matmat();
   if (succeeded(matmatRes)) {
-    rewriter.replaceOp(rootOp, *matmatRes);
-    return success();
+    return matmatRes;
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:699]: Why don't we replace the root op here? That should preserve the same structure as for other patterns

**Review by dcaballe — CHANGES_REQUESTED** [review]
LG! A few comments

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp (hunk 1 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp]: > This is the main concern I've been struggling with for a while: this pattern matches the nested op and the rewrite would replace its parent. I'm not quite sure pattern rewrite allows that even though it works for this cases

I wasn't sure myself so I asked:

* https://discourse.llvm.org/t/oprewritepattern-updating-the-parent-op/

I guess that this should be fine then?

> I couldn't find an elegant way to provide this functionality but maybe you can :)

No better ideas just yet :/

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp (hunk 5 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp:713]: We have two cases though:
* `vector.contract` --> here we should replace `vector.contract`
* `vector.mask {vector.contract}` --> here we should replace `vector.mask` (no need to "replace" `vector contract`)

This **can be** implemented, but then you will require logic like this in every pattern:
```cpp
  OpBuilder::InsertionGuard guard(rewriter);
  auto maskableOp = cast<vector::MaskableOpInterface>(op.getOperation());
  Operation *rootOp;
  if (maskableOp.isMasked()) {
    rewriter.setInsertionPoint(maskableOp.getMaskingOp());
    rootOp = maskableOp.getMaskingOp();
  } else {
    rootOp = op;
  }
```

That's something that I am removing and instead moving to the base class, `MaskableOpRewritePattern`. But it also means I can't "replace" anymore as the logic to decide "what" to replace has been removed.

Am I overthinking this? 🤔

File: mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp (hunk 4 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/Transforms/LowerVectorContract.cpp]: Done!

**Review by dcaballe — APPROVED** [review]
Thanks!

**Pull request merged**